### ISSUE TITLE
Status & Headers in HttpResponse not transferred to MicronautAwsProxyResponse

### ISSUE THREAD

**Issue by thobias-karlsson-wcar** [micronaut-projects/micronaut-aws#1387]
### Expected Behavior

When `HttpResponse<?>` is returned from an `ExceptionHandler<T, HttpResponse<R>>` with a non-200 status code and custom headers I expect the `MicronautAwsProxyResponse` that is returned to AWS API Gateway to contain the same. 

**UPDATE**: This bug also applies to normal `@Controller` methods. The following code will end up with `200 OK` in AWS instead of expected `202 Accepted`.
``` 
@Controller("/test")
public class TestController {

  @Post("/custom-status")
  public MutableHttpResponse<?> customStatus() {
    return new SimpleHttpResponseFactory().status(HttpStatus.ACCEPTED);
  }
}
```
However, using `@Status(HttpStatus.ACCEPTED)` works as expected. This is however not a solution when dealing with error codes afaik.

### Actual Behaviour

The `HttpResponse.status` is not transferred into the resulting AWS proxy response, hence the response will always be a `200 OK`. The same thing applies to headers. 

### Steps To Reproduce

1. Implement an exception handler bean that maps an exception to a custom HTTP error status code, e.g:
```
  @Produces
  @Singleton
  public static class HttpExceptionHandler
      implements ExceptionHandler<HttpException, HttpResponse<ProblemDto>> {

    private static final String CONTENT_TYPE_PROBLEM = "application/problem+json";

    @Override
    public HttpResponse<ProblemDto> handle(HttpRequest request, HttpException exception) {
      ProblemDto problemDto = exception.getProblemDto();
      problemDto.instance(request.getUri());
      if (isClientError(exception.getStatus())) {
        if (LOG.isDebugEnabled()) {
          LOG.debug("Client Error caught\n{}", describeWithTrace(exception));
        }
      } else {
        LOG.error("Server Error caught", exception);
      }
      return new SimpleHttpResponseFactory().status(exception.getStatus())
          .contentType(CONTENT_TYPE_PROBLEM)
          .body(problemDto);
    }
  }
```
2. When running unit tests using `@MicronautTest` and enforcing the exception, you get the expected status code along with the expected custom headers. _But_ when running the application as a `MicronautLambdaHandler` in AWS, the lambda will always end up with `200 OK`. This can be reproduced with unit testing by initializing the `MicronautLambdaHandler` the exact same way as AWS Lambda will initialize it, i.e. using default constructor. Example test:
```
class MicronautLambdaHandlerTest {

  MicronautLambdaHandler unitUnderTest;

  @BeforeEach
  void beforeEach() throws ContainerInitializationException {
    unitUnderTest = new MicronautLambdaHandler();
  }

  @AfterEach
  void afterEach() {
    unitUnderTest.close();
  }

  @Test
  void test() {
    AwsProxyRequest input = ...

    AwsProxyResponse response = unitUnderTest.handleRequest(input, null);

    // Then
    assertThat(response).isNotNull();
    assertThat(response.getStatusCode()).isEqualTo(400);
  }
}
```
3. This test fails due to status being 200 OK. 
4. When debugging the test I end up on this line, where only the `body` is transferred to the resulting AWS Proxy Response.
https://github.com/micronaut-projects/micronaut-aws/blob/9efab634cd39e804b0391c80a0939e2ce3befc66/function-aws-api-proxy/src/main/java/io/micronaut/function/aws/proxy/MicronautLambdaContainerHandler.java#L348

TL;DR. We need to be able to dynamically set HTTP status code & headers (without annotation magic).

### Environment Information

Micronaut BOM `3.4.4` which points to `micronaut-function-aws-api-proxy` version `3.2.3`.

### Version

3.4.4

**Comment by thobias-karlsson-wcar** [comment]
Here is an example unit test class with the same HTTP request but using `@MicronautTest` and the Micronaut `HttpClient`. This test is successful and returns the expected status code (`400`). 

```
@MicronautTest
class MicronautControllerTest {

  @Inject
  @Client("/")
  HttpClient client;

  @Test
  void test() {
    // Given
    MutableHttpRequest<Object> request = ...

    // When
    HttpResponse<Object> response = client.toBlocking().exchange(request);

    // Then
    assertThat(response.getStatus().getCode()).isEqualTo(400);
  }
}
```
Our entire application is tested using `@MicronautTest` with several 100 controller tests verifying status codes, response headers etc. We noticed the `200 OK` issue during integration tests against live AWS environment, so I would really like to get this fixed - or a proposed solution to what we are doing wrong (?). To me it's very strange behaviour by Micronaut.

**Comment by thobias-karlsson-wcar** [comment]
I found the following in the documentation which would support that this is actually a bug, right?

> 6.8 Response Status
> A Micronaut’s controller action responds with a 200 HTTP status code by default.
> If the controller’s action returns a HttpResponse object, you can configure the status code for the response object with the status method.
Ref: https://docs.micronaut.io/2.0.0.M2/guide/index.html#statusAnnotation

**Comment by thobias-karlsson-wcar** [comment]
I tried to override `MicronautLambdaHandler.handleRequest(AwsProxyRequest input, Context context)` (see below). Now it works, but it's a very temporary hack and it only works because our application returns a Problem JSON according to RFC 7807 which includes the HTTP status in root `status` field. We have other endpoints which should return non-200 and non-client-error codes like `202 Accepted` which still doesn't work along with the response headers that we want to return (according to W3C Trace Context). Please help us out. :)

Temporary override:
```
public class RestApiHandler extends MicronautLambdaHandler {

  ...

  @Override
  public AwsProxyResponse handleRequest(AwsProxyRequest input, Context context) {
    final AwsProxyResponse proxyResponse = super.handleRequest(input, context);
    final String responseBody = proxyResponse.getBody();
    if (proxyResponse.getStatusCode() == 200 && responseBody != null) {
      getProblemStatus(responseBody)
          .ifPresent(problemStatus -> {
            LOG.warn("Replacing HTTP status code {} with {}", proxyResponse.getStatusCode(), problemStatus);
            proxyResponse.setStatusCode(problemStatus.getCode());
            proxyResponse.setStatusDescription(problemStatus.getReason());
            proxyResponse.addHeader("Content-Type", "application/problem+json");
          });
    }
    return proxyResponse;
  }

  private Optional<HttpStatus> getProblemStatus(@NonNull String body) {
    try {
      return Optional.ofNullable(OBJECT_MAPPER.readTree(body))
          .filter(node -> node.has("title"))
          .filter(node -> node.has("instance"))
          .filter(node -> node.has("status"))
          .map(node -> node.get("status"))
          .map(JsonNode::asInt)
          .map(HttpStatus::valueOf);
    } catch (JsonProcessingException e) {
      LOG.warn("Unable to look for 'status' in body: {}", body, e);
      return Optional.empty();
    }
  }
}
```

**Comment by thobiaskarlsson** [comment]
This bug is also present in Micronaut 3.5.0.

**Comment by sdelamo** [comment]
I have started a PR but I have not being able to reproduce so far: 

https://github.com/micronaut-projects/micronaut-aws/pull/1392

**Comment by sdelamo** [comment]
can you point me to what I am doing wrong in my PR to reproduce it or a link to a sample project which reproduces the issue?

**Comment by thobiaskarlsson** [comment]
Your tests seem fine, which amazes me since it obviously doesn't work for us. Let me get back to you with an example application.

**Comment by thobiaskarlsson** [comment]
@sdelamo I've now created a very minimal version of our application including same dependencies and versions and successfully reproduced the bug (see repo below). There are two test classes:
- `MyControllerMicronautTest` that contains three test cases built on the `@MicronautTest` annotation and an `@Inject`ed Micronaut `HttpClient`. All test cases are successful here.
- `MyControllerLambdaHandlerTest` that contains the same three test cases as in previous class, but instead of using `@MicronautTest` annotation and client calls the `MicronautLambdaHandler` directly with an `AwsProxyRequest`. Here 2/3 test cases are failing. The third test case which is successul requests an endpoint in my `@Controller`which is annotated with `@Status(HttpStatus.ACCEPTED)` while the other (failing tests) are hitting endpoints that return `MutableHttpResponse<?>` and `HttpResponse<?>`.
- 
Evidence repo: https://github.com/thobiaskarlsson/micronaut-aws-1387

In our real repository we initialize the AWS Lambda using CloudFormation (re-written to match naming in tmp repo):
```
  ApplicationLambda:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: ApplicationLambda
      Handler: se.thoka.tmp.MyLambdaHandler
      ...
```

**Comment by sdelamo** [comment]
Instead of :
```
return new SimpleHttpResponseFactory()
        .status(HttpStatus.ACCEPTED)
        .body("Testing getStatusMutableHttpResponse");
```

you should use: 

```java
return HttpResponse.status(HttpStatus.ACCEPTED)
        .body("Testing getStatusMutableHttpResponse");
```

**Comment by sdelamo** [comment]
You should use the fluid api to create requests. 

Insted of: 

```java
    MutableHttpRequest<Object> request = HttpRequest.create(HttpMethod.GET, "/test/status-annotated");
```
use:
```java
HttpRequest<?> request = HttpRequest.GET("/test/status-annotated").accept(MediaType.TEXT_PLAIN);
```.

Please, note that if your controller does not respond JSON, add the appropriate accept header.

**Comment by thobiaskarlsson** [comment]
Sure, I can fix that, but the issue still remains, this shouldn't affect the outcome of the test afaik. Please note that I used HttpResponse.status(...) in one of the endpoints and MutableHttpResponse in the other method just to show you that both fails. Reason for this is that somebody imposed that I need to use MutableHttpResponse instead of HttpResponse for it to work, but they end up with the exact same result. 

Regarding the HttpRequest.GET, this is neither something that will affect the outcome.

**Comment by sdelamo** [comment]
@thobias-karlsson-wcar Let me know if this helps.

**Comment by thobiaskarlsson** [comment]
You've removed a lot of dependencies from the POM, hence I can't even get your PR to build. :thinking: I also don't see the point of introducing the `Application` class. What am I missing here? 

Should `MyLambdaHandler` still be my entrypoint for the AWS Lambda? You didn't do any changes to that class and since you removed the `micronaut-runtime` dependency and the class extends `MicronautLambdaHandler` (as per Micronaut documentation) it does not build (neither `com.amazonaws.serverless.exceptions.ContainerInitializationException` nor `io.micronaut.function.aws.proxy.MicronautLambdaHandler` exists on classpath).

**Comment by thobiaskarlsson** [comment]
Sorry, I added a comment and deleted it again, I missed one thing. Now it's building fine. :)

So. You are correct, your PR works fine, but I don't understand _why_ and _how_ it's even possible. You've removed _a lot_ of Micronaut dependencies, added an `Application` class that does not seem to be used anywhere (the `<exec.mainClass>com.example.Application</exec.mainClass>` in the POM refers to a non-existing package/class). , which is why I simply don't understand how this could work. :thinking: Is it the `micronaut-maben-plugin` and the `maven-compiler-plugin` configuration that does all the magic? I will have to try to apply your changes to our application and see if it works there as well. There is sooo much magic going on here imho, but I'll give it a shot. Thanks!

**Comment by thobiaskarlsson** [comment]
Ok, I stand corrected. This is so weird. I replicated most of your changes into our real application and now it seems to be working. :+1: What I did:

- Added `<micronaut.version>3.5.0</micronaut.version>` property to POM
- Removed superfluous Micronaut dependencies
- Changed annotation processor paths from 
``` 
                    <annotationProcessorPaths>
                        <path>
                            <groupId>io.micronaut</groupId>
                            <artifactId>micronaut-inject-java</artifactId>
                            <version>${micronaut.version}</version>
                        </path>
                        <path>
                            <groupId>org.projectlombok</groupId>
                            <artifactId>lombok</artifactId>
                            <version>${lombok.version}</version>
                        </path>
                    </annotationProcessorPaths>
```
into 
``` 
                    <annotationProcessorPaths combine.children="append">
                        <path>
                            <groupId>org.projectlombok</groupId>
                            <artifactId>lombok</artifactId>
                            <version>${lombok.version}</version>
                        </path>
                    </annotationProcessorPaths>
``` 

And that's it. What could possibly have been the issue here? Please note that I skipped adding the `Application` and `<exec.mainClass>` property, still works. I'm VERY happy, but equally confused. :) Thank you very much @sdelamo for helping us out. Kudos!

**Comment by thobiaskarlsson** [comment]
@sdelamo I can now inform you that this actually worked. Application successfully deployed to AWS and the correct status codes are being returned in response. :pray: Thanks!
